### DISCUSSION TITLE
Local vs remote execution

### DISCUSSION THREAD

**Discussion by teknowledgist** [sumatrapdfreader/sumatrapdf#2351] [Q&A] [answered]
I have a PowerShell script that builds a PDF and silently prints it via SumatraPDF.  It works great if I RDP to the computer and run it in a PowerShell console.  The line that does the actual printing is:

    Start-Process -FilePath $SumatraPath -ArgumentList "-Print-To $Printer `"$File`"" 

Pretty straightforward.  However, I need to fire this off from a webserver via an OpenSSH connection, and something is causing SumatraPDF to hang.  The same exact script and/or manual command while connected via OpenSSH as the same user and... nothing.  The SumatraPDF process starts, but nothing ever prints and the process never closes which means there is no error code.  Since this is a command-line-only interface, I can't open a debugging window or see any potential dialog box that SumatraPDF is trying to present.  

Although the testing account and the one that will be using SSH are unprivileged, the same hang occurs when this is attempted as an administrator.

Is there a log somewhere that might give me a better idea of what the hangup is?

Thanks.

**Comment by GitHubRulesOK** [comment]
Generally SumatraPDF in common with Acrobat Reader is not designed for headless/remote printing since such configurations dont present the same characteristics as a WorkStation. Notoriously Windows printing uses GDI so has been borked several times by MS over the past year as they tighten application threading access. You will need to check how your application relates to UAC and GDI so look on the server screen to see the error message(s) and if it has no screen thats possibly the answer..

**Reply by teknowledgist** [reply]
I'm trying to use SumatraPDF specifically because it has silent, command-line switches for printing.  Are those options "experimental" or depreciated and not really expected to work in some situations?

I don't have an "application" to relate to UAC or GDI.  It's just a PowerShell script that tells SumatraPDF to do something.  The failure is something within SumatraPDF, and I can't work around it if I can't see the cause.

Is there no way to generate a debug log for SumatraPDF?

**Reply by kjk** [reply]
Sumatra wasn't designed or tested for such scenarios and there's no logging for that.
It's an open-source app so you can add logging or even run it under the debugger from your script to see where exactly it fails.

**Reply by teknowledgist** [reply]
Ugh.  OK.  Thanks for the follow up.  I guess it's back to the drawing board.

**Reply by GitHubRulesOK** [reply]
@teknowledgist The application designed for printing many formats to devices is GhostScript and I have used that for remote service printing  so it may be worth looking closer at that, since it is not dependant on a GDI platform as its multi-platform.

**Reply by teknowledgist** [reply]
Just a follow up.  I tried GhostScript, and was not able to get it to work.  My working solution was to skip the directed printing, but instead just generate a PDF and use [AutoPrint](https://www.4-tech-engineering.com/autoprint.html) as a service to watch for files in a particular location and send them to a particular printer.  I need to give props to the support -- they were super-responsive and helped through lots and lots of free back-and forth for me to figure out how to get it working for my needs.

**Reply by GitHubRulesOK** [reply]
Interesting they are using an Adobe/Google/Amazon/Microsoft  collaborator https://www.foxit.com/quick-pdf-library/

**Reply by teknowledgist** [reply]
It wouldn't have been my first choice, but the number of promising paths with unexpected dead-ends I have encountered on this "simple" project (including the subject of this thread) is amazing to me.  Ultimately, I just needed a solution that worked.

**Comment by teknowledgist** [comment, marked as answer]
Another follow up because I figured it out!  

When I changed to secure-key authentication for my OpenSSH connection, I needed to make some additions to my script because of the "second-hop problem".  [See here](https://superuser.com/questions/1710006/how-to-query-ad-within-ssh-key-based-authentication-session/1710016#1710016)  Anyway, I thought just for giggles that I would try re-authenticating for the SumatraPDF print process:

    $NetID = "myDomain\username"
    [pscredential]$credObject = New-Object System.Management.Automation.PSCredential ($NetID, $secStringPassword)
    Start-Process -FilePath $SumatraPath -ArgumentList "-Print-To $Printer `"$File`"" -credential $credObject

and that works.  Now, I can have direct control of printing (via SumatraPDF) rather than relying on another service to "catch" my PDFs.

Ultimately, this was not a SumatraPDF problem at all.  Rather, it was totally just a quirk in how Windows handles authentication.  I have no idea why password authentication allowed other "continued" authentication (i.e. querying AD), but not for printing.  

I hope this helps someone, someday, and thanks for hearing me out.
